ReadMe
wifimgr WiFi Networks Manager

The latest release tarball is available here:
	http://opal.com/freebsd/ports/net-mgmt/wifimgr

The current development version can be downloaded using subversion here:
	svn://opal.com/freebsd/ports/net-mgmt/wifimgr


ChangeLog
2013/05/27 fbsd@opal.com
	Version 1.11
	- Added support for WiFi interface ifconfig in /etc/rc.conf.local.
	- Remove exit() when failing to open an /etc/rc.conf* file.
	- Updated .pot file.
	- WITHOUT_NLS support fixes
	- fixes to keep clang(1) happy

2012/01/28 fbsd@opal.com
	Version 1.10
	- Added Russion translation, thanks to Serge Volkov <sv@sisadminov.net>.
	- Use uid(0) rather than nam("root") for super-user authentication.

2011/11/15 fbsd@opal.com
	Version 1.9
	- Allow 64 character hexadecimal PSK keys.
	- Minor corrections to code indentation.
	- Added Bulgarian translation, thanks to Любомир Григоров <nm.knife@gmail.com>.
	- Correction to German translation.
	- Msgmerge update of all translations.
	- Added masking of password input and also a checkbox to un-mask the password.
	- Added checkbox and meta-config to not save bssid, e.g., for large campus
	  networks or mobile phone hotspots.  And made changes to associated network code
	  to work both with bssid configured and without, when it has to match
	  on ssid.
	- Fixed logic error regarding security type for EAP networks.

2011/08/15 fbsd@opal.com
	Version 1.8
	- added Japanese translation, thanks to TAKANO Yuji (たかのゆ〜じ) <takachan@running-dog.net>.
	- added Dutch translation, done myself.
	- added German translation, thanks to Frank Ebert <frank.ebert@freenet.de>.

2011/01/30 fbsd@opal.com
	Version 1.7
	- added ability to manually add networks, e.g., cloaked networks
	- do not display networks from scan when interface is down
	- fix up/down status if Rescan or Save/Rescan clicked while i/f down
	- use bssid rather than ssid to determine where to display green associated icon
	- after interface restart, poll for 10 seconds until it re-associates

2010/08/31 fbsd@opal.com
	Version 1.6
	- added support for WiFi interface cloned under lagg(4) interface
	- minor change to signal strength (bars) formula

2009/12/29 fbsd@opal.com
	Version 1.5
	- added support for RSN network announcements
	- added support for EAP authentication
	- detect when network's security type changed and alert user
	- replace fscanf() calls with fgets() to eliminate a hang some folk reported
	- added extra info on WiFi security to man page
	- added WiFi Up/Down button, on-off.png and wifimgr-grey.png icons
	- create empty wpa_supplicant.conf file if it doesn't exist
	- change to ifconfig_associated_network() to properly parse quoted SSID

2009/12/18 fbsd@opal.com
	Version 1.4
	- added support for TMPDIR environment variable
	- removed DEBUG printf() left in previous version
	- corrections to buf pointers in several [fs]scanf() calls
	- replaced two sscanf() calls with simpler strcpy() calls
	- added proper #include for umask() definition
	- added "return 1" at end of a non-void function
	- removed an unused variable

2009/12/13 fbsd@opal.com
	Version 1.3
	- use BSSID to differentiate networks with same SSID
	  (requires multiple entries for networks with multiple access points)
	- removed an unused function

2009/12/03 fbsd@opal.com
	Version 1.2
	- French translation, thanks to Patrick Lamaizière <patfbsd@davenulle.org>.
	- added call to bind_textdomain_codeset() in NLS initialization.
	- added NLS initialization to wifimgrsu.c also.
	- updated wifimgr.pot template for current line numbering.

2009/11/24 fbsd@opal.com
	Version 1.1
	- set PATH to known value (needed for /etc/rc.d/netif)
	- clear gui_changed flag on saving

2009/11/23 fbsd@opal.com
	Version 1.0.2
	- clarification to LICENSE-psdGraphics.com that our use of the two
	  icons is acceptable to them

2009/11/15 fbsd@opal.com
	Version 1.0.1
	- change to src/Makefile to create share/applications dir, if needed
	- change to FreeBSD port pkg_plist to remove share/applications dir
	- change to FreeBSD port Makefile to fix NLS comments

2009/11/10 fbsd@opal.com
	Version 1.0 released


src/string.c
/*
** string.c
*/

/*
** Copyright (c) 2009, J.R. Oldroyd, Open Advisors Limited
** All rights reserved.
** 
** Redistribution and use in source and binary forms, with or without
** modification, are permitted provided that the following conditions are met:
**     * Redistributions of source code must retain the above copyright
**       notice, this list of conditions and the following disclaimer.
**     * Redistributions in binary form must reproduce the above copyright
**       notice, this list of conditions and the following disclaimer in the
**       documentation and/or other materials provided with the distribution.
**     * Neither the name of the author, the author's organization nor the
**       names of its contributors may be used to endorse or promote products
**       derived from this software without specific prior written permission.
** 
** THIS SOFTWARE IS PROVIDED BY OPEN ADVISORS LIMITED ''AS IS'' AND ANY
** EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
** WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
** DISCLAIMED. IN NO EVENT SHALL OPEN ADVISORS LIMITED BE LIABLE FOR ANY
** DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
** (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
** LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
** ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
** (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
** SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

/*
** $Id: string.c 60 2009-12-23 19:11:11Z jr $
*/

#include <string.h>

/*
** remove trailing newline from string
*/
int
chop(char * s) {
	char *			p;

	if (*s) {
		p = s + strlen(s) - 1;
		if (*p == '\n') {
			*p = '\0';
			return 1;
		}
	}
	return 0;
}


src/version.h
/*
** version.h
*/

/*
** Copyright (c) 2009, J.R. Oldroyd, Open Advisors Limited
** All rights reserved.
** 
** Redistribution and use in source and binary forms, with or without
** modification, are permitted provided that the following conditions are met:
**     * Redistributions of source code must retain the above copyright
**       notice, this list of conditions and the following disclaimer.
**     * Redistributions in binary form must reproduce the above copyright
**       notice, this list of conditions and the following disclaimer in the
**       documentation and/or other materials provided with the distribution.
**     * Neither the name of the author, the author's organization nor the
**       names of its contributors may be used to endorse or promote products
**       derived from this software without specific prior written permission.
** 
** THIS SOFTWARE IS PROVIDED BY OPEN ADVISORS LIMITED ''AS IS'' AND ANY
** EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
** WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
** DISCLAIMED. IN NO EVENT SHALL OPEN ADVISORS LIMITED BE LIABLE FOR ANY
** DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
** (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
** LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
** ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
** (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
** SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

/*
** $Id: version.h 99 2013-05-27 20:20:35Z jr $
*/

#define VERSION		"Version 1.11"


src/wifimgr.h
/*
** wifimgr.h
*/

/*
** Copyright (c) 2009, J.R. Oldroyd, Open Advisors Limited
** All rights reserved.
** 
** Redistribution and use in source and binary forms, with or without
** modification, are permitted provided that the following conditions are met:
**     * Redistributions of source code must retain the above copyright
**       notice, this list of conditions and the following disclaimer.
**     * Redistributions in binary form must reproduce the above copyright
**       notice, this list of conditions and the following disclaimer in the
**       documentation and/or other materials provided with the distribution.
**     * Neither the name of the author, the author's organization nor the
**       names of its contributors may be used to endorse or promote products
**       derived from this software without specific prior written permission.
** 
** THIS SOFTWARE IS PROVIDED BY OPEN ADVISORS LIMITED ''AS IS'' AND ANY
** EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
** WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
** DISCLAIMED. IN NO EVENT SHALL OPEN ADVISORS LIMITED BE LIABLE FOR ANY
** DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
** (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
** LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
** ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
** (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
** SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

/*
** $Id: wifimgr.h 102 2014-05-10 20:53:58Z jr $
*/

#define NETWORKS_FILE	"/etc/wpa_supplicant.conf"
#define RC_CONF_FILE	"/etc/rc.conf"
#define RC_CONF_LOCAL_FILE	"/etc/rc.conf.local"

#define ICON_PATH	"/usr/local/share/wifimgr/icons"

#define PATH_CAT	"/bin/cat"
#define PATH_CP		"/bin/cp"
#define PATH_DIFF	"/usr/bin/diff"
#define PATH_IFCONFIG	"/sbin/ifconfig"
#define PATH_NETIF	"/etc/rc.d/netif"
#define PATH_SED	"/usr/bin/sed"
#define PATH_WIFIMGRSU	"/usr/local/libexec/wifimgrsu"

#define ETCDIR		"/etc"
#define TMPDIR		"/tmp"

struct wifi_net {
	int			wn_enabled;
	unsigned int		wn_security;
#define WN_SEC_UNKNOWN		0x0
#define WN_SEC_NONE		0x1
#define WN_SEC_WEP		0x2
#define WN_SEC_WPA		0x4
#define WN_SEC_RSN		0x8
	unsigned int		wn_km;
#define WN_KM_UNKNOWN		0x0
#define WN_KM_NONE		0x1
#define WN_KM_PSK		0x2
#define WN_KM_EAP		0x4
	char			wn_key[256];
	char			wn_comment[256];
	int			wn_any_bssid;
	int			wn_chan;
	int			wn_rate;
	int			wn_bars;
	struct wifi_net *	wn_next;

	int			wn_show_password;
	/* next four are really GtkWidget * pointers */
	void *			w_wn_key;
	void *			w_sup_password;
	void *			w_sup_private_key_passwd;
	void *			w_sup_private_key2_passwd;

	char			sup_anonymous_identity[256];
	char			sup_auth_alg[256];
	char			sup_bssid[256];
	char			sup_ca_cert[256];
	char			sup_ca_cert2[256];
	char			sup_client_cert[256];
	char			sup_client_cert2[256];
	char			sup_dh_file[256];
	char			sup_dh_file2[256];
	char			sup_eap[256];
	int			sup_eapol_flags;
	char			sup_eappsk[256];
	int			sup_eap_workaround;
	char			sup_group[256];
	char			sup_identity[256];
	char			sup_key_mgmt[256];
	int			sup_mixed_cell;
	int			sup_mode;
	char			sup_nai[256];
	char			sup_pac_file[256];
	char			sup_pairwise[256];
	char			sup_password[256];
	char			sup_phase1[256];
	char			sup_phase2[256];
	int			sup_priority;
	char			sup_private_key[256];
	char			sup_private_key2[256];
	char			sup_private_key_passwd[256];
	char			sup_private_key2_passwd[256];
	char			sup_proto[256];
	char			sup_psk[256];
	int			sup_scan_ssid;
	char			sup_server_nai[256];
	char			sup_subject_match[256];
	char			sup_subject_match2[256];
	char			sup_ssid[256];
	char			sup_wep_key0[256];
	char			sup_wep_key1[256];
	char			sup_wep_key2[256];
	char			sup_wep_key3[256];
	int			sup_wep_tx_keyidx;
};

struct conflist {
	char *			cl_name;
	int			cl_value;
};

struct conflist			conf_auth_alg[4];
struct conflist			conf_eap[8];
struct conflist			conf_group[5];
struct conflist			conf_key_mgmt[4];
struct conflist			conf_pairwise[4];
struct conflist			conf_proto[4];
extern struct wifi_net *	nets;
extern char *			wifi_if;
extern int			wifi_if_status;
#define WIFI_IF_DOWN		0
#define WIFI_IF_UP		1

extern int			toggle_intf_up_down(char *);
extern int			ifconfig_intf_status(char *);
extern char *			ifconfig_associated_network(char *);
extern struct wifi_net *	build_network_list(char *);
extern int			save_networks_file(char *);
extern int			restart_intf(void);

extern int			chop(char * s);

extern int			nl_insert(struct wifi_net **, struct wifi_net *);
#define NL_INSERT_ERROR		0
#define NL_INSERT_OK		1
#define NL_INSERT_SEC_CHG	2
extern int			nl_delete_list(struct wifi_net **);

extern void			nl_order_by_ssid(struct wifi_net **);
extern void			nl_order_by_bars(struct wifi_net **);
extern void			nl_order_by_channel(struct wifi_net **);

extern int			gui_changes;
extern char			gui_response[];
extern int			gui_init();
extern int			gui_message(const char *, int);
#define	MSG_INFO		0
#define	MSG_WARNING		1
#define	MSG_QUESTION		2
#define	MSG_ERROR		3
#define	MSG_INPUT		4
#define MSG_RESPONSE_OK		0
#define MSG_RESPONSE_FAIL	1
#define MSG_RESPONSE_YES	2
#define MSG_RESPONSE_NO		3
extern void			gui_loop();


src/net_list.c
/*
** net_list.c
*/

/*
** Copyright (c) 2009, J.R. Oldroyd, Open Advisors Limited
** All rights reserved.
** 
** Redistribution and use in source and binary forms, with or without
** modification, are permitted provided that the following conditions are met:
**     * Redistributions of source code must retain the above copyright
**       notice, this list of conditions and the following disclaimer.
**     * Redistributions in binary form must reproduce the above copyright
**       notice, this list of conditions and the following disclaimer in the
**       documentation and/or other materials provided with the distribution.
**     * Neither the name of the author, the author's organization nor the
**       names of its contributors may be used to endorse or promote products
**       derived from this software without specific prior written permission.
** 
** THIS SOFTWARE IS PROVIDED BY OPEN ADVISORS LIMITED ''AS IS'' AND ANY
** EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
** WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
** DISCLAIMED. IN NO EVENT SHALL OPEN ADVISORS LIMITED BE LIABLE FOR ANY
** DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
** (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
** LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
** ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
** (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
** SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

/*
** $Id: net_list.c 88 2011-11-15 15:48:26Z jr $
*/

#include "wifimgr.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

static int (*_nl_sort_specific_compar)(struct wifi_net *, struct wifi_net *);

static int
_nl_sort_generic_compar(const void * a, const void * b) {
	return _nl_sort_specific_compar(*(struct wifi_net **)a, *(struct wifi_net **)b);
}

static void
_nl_sort(struct wifi_net ** a, int (*compar)(struct wifi_net *, struct wifi_net *)) {
	struct wifi_net *	p;
	struct wifi_net **	list;
	size_t			num;
	int			i;

	for (p = *a, num = 0; p; p = p->wn_next)
		num++;

	if (num <= 1)
		return;

	list = malloc(num * sizeof(struct wifi_net *));
	if (list == NULL)
		return;

	for (p = *a, i = 0; p; p = p->wn_next, i++)
		list[i] = p;

	_nl_sort_specific_compar = compar;
	qsort(list, num, sizeof(struct wifi_net *), _nl_sort_generic_compar);

	*a = p = list[0];
	for (i = 1; i < num; i++) {
		p->wn_next = list[i];
		p = list[i];
	}
	p->wn_next = NULL;

	free(list);
}

static int
_nl_order_by_ssid_compar(struct wifi_net * a, struct wifi_net * b) {
	int cmp = strcasecmp(a->sup_ssid, b->sup_ssid);
	if (cmp != 0)
		return cmp;

	return strcasecmp(a->sup_bssid, b->sup_bssid);
}

void
nl_order_by_ssid(struct wifi_net ** a) {
	_nl_sort(a, _nl_order_by_ssid_compar);
}

static int
_nl_order_by_bars_compar(struct wifi_net * a, struct wifi_net * b) {
	if (a->wn_bars == b->wn_bars) {
		int cmp = strcasecmp(a->sup_ssid, b->sup_ssid);
		if (cmp != 0)
			return cmp;

		return strcasecmp(a->sup_bssid, b->sup_bssid);
	}

	return b->wn_bars - a->wn_bars;
}

void
nl_order_by_bars(struct wifi_net ** a) {
	_nl_sort(a, _nl_order_by_bars_compar);
}


static int
_nl_order_by_channel_compar(struct wifi_net * a, struct wifi_net * b) {
	if (a->wn_chan == b->wn_chan) {
		int cmp = strcasecmp(a->sup_ssid, b->sup_ssid);
		if (cmp != 0)
			return cmp;

		return strcasecmp(a->sup_bssid, b->sup_bssid);
	}

	/* 0 channel is always last */
	if (a->wn_chan == 0)
		return 1;
	if (b->wn_chan == 0)
		return -1;

	return a->wn_chan - b->wn_chan;
}

void
nl_order_by_channel(struct wifi_net ** a) {
	_nl_sort(a, _nl_order_by_channel_compar);
}

/*
** copy and insert element into list
*/
int
nl_insert(struct wifi_net ** a, struct wifi_net * b) {
	struct wifi_net *	p;
	struct wifi_net *	pp;
	int			security_changed;

	if (!a || !b)
		return NL_INSERT_ERROR;

	/* search for this or next alphabetically higher SSID */
	for (p = *a; p; pp = p, p = p->wn_next)
		if (strcasecmp(p->sup_ssid, b->sup_ssid) >= 0)
			break;

	/* search within SSID for this or empty or next alphabetically higher BSSID */
	for ( ; p; pp = p, p = p->wn_next) {
		if (strcasecmp(p->sup_ssid, b->sup_ssid) != 0)
			break;
		if (strcasecmp(p->sup_bssid, b->sup_bssid) >= 0 || strcmp(p->sup_bssid, "") == 0)
			break;
	}

	security_changed = 0;
	if (p && strcasecmp(p->sup_ssid, b->sup_ssid) == 0 &&
	    (strcasecmp(p->sup_bssid, b->sup_bssid) == 0 ||
	    strcmp(p->sup_bssid, "") == 0 ||
	    p->wn_any_bssid)) {
		/* net is already in list, so merge */
		if (strcmp(p->sup_bssid, "") == 0 && !p->wn_any_bssid)
			strcpy(p->sup_bssid, b->sup_bssid);
		p->wn_chan = b->wn_chan;
		p->wn_rate = b->wn_rate;
		p->wn_bars = b->wn_bars;
		if (! p->wn_security || ! *p->sup_proto) {
			p->wn_security = b->wn_security;
			strcpy(p->sup_proto, b->sup_proto);
		}
		else if (p->wn_security & b->wn_security) {
			/* b is a subset of p, so just reduce p */
			p->wn_security = b->wn_security;
			strcpy(p->sup_proto, b->sup_proto);
		}
		else if (p->wn_security != b->wn_security) {
			p->wn_security = b->wn_security;
			strcpy(p->sup_proto, b->sup_proto);
			security_changed++;
		}
		if (! p->wn_km)
			p->wn_km = b->wn_km;
		else if (p->wn_km != b->wn_km) {
			p->wn_km = b->wn_km;
			security_changed++;
		}
	}
	else {
		/* insert */
		if (p == *a) {
			b->wn_next = *a;
			*a = b;
		}
		else {
			b->wn_next = p;
			pp->wn_next = b;
		}
	}

	return (security_changed) ? NL_INSERT_SEC_CHG : NL_INSERT_OK;
}


/*
** delete all items in list
*/
int
nl_delete_list(struct wifi_net ** a) {
	struct wifi_net *	p;
	struct wifi_net *	pp = NULL;

	if (!a)
		return 0;

	for (p = *a; p; pp = p, p = p->wn_next)
		if (pp)
			free(pp);
	if (p)
		free(p);

	*a = NULL;
	return 1;
}


src/wifimgrsu.c
/*
** wifimgrsu.c
**	execute command as super-user on behalf of wifimgr(8)
*/

/*
** Copyright (c) 2009, J.R. Oldroyd, Open Advisors Limited
** All rights reserved.
** 
** Redistribution and use in source and binary forms, with or without
** modification, are permitted provided that the following conditions are met:
**     * Redistributions of source code must retain the above copyright
**       notice, this list of conditions and the following disclaimer.
**     * Redistributions in binary form must reproduce the above copyright
**       notice, this list of conditions and the following disclaimer in the
**       documentation and/or other materials provided with the distribution.
**     * Neither the name of the author, the author's organization nor the
**       names of its contributors may be used to endorse or promote products
**       derived from this software without specific prior written permission.
** 
** THIS SOFTWARE IS PROVIDED BY OPEN ADVISORS LIMITED ''AS IS'' AND ANY
** EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
** WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
** DISCLAIMED. IN NO EVENT SHALL OPEN ADVISORS LIMITED BE LIABLE FOR ANY
** DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
** (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
** LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
** ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
** (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
** SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

/*
** $Id: wifimgrsu.c 102 2014-05-10 20:53:58Z jr $
*/

#include "wifimgr.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <unistd.h>
#include <pwd.h>

#ifndef WITHOUT_NLS
#include <libintl.h>
#else
#define gettext(x)	(x)
#endif

#define	PATH		"/bin:/usr/bin:/sbin:/usr/sbin"

int
main(int argc, char ** argv) {
	char			line[1024];
	char			cmd[1024];
	int			auth = 0;
	int			copy_stdout;
	char *			arg;
	int			xst;

#ifndef WITHOUT_NLS
	/* set output character set */
	bind_textdomain_codeset("wifimgr", "UTF-8");

	/* set language catalog domain */
	textdomain("wifimgr");
#endif

	if (isatty(0) || isatty(1)) {
		fprintf(stderr, gettext("wifimgrsu: for invocation by wifimgr(8) only\n"));
		exit(1);
	}

	/* set line buffering */
	setvbuf(stdin, (char *) NULL, _IOLBF, 0);
	setvbuf(stdout, (char *) NULL, _IOLBF, 0);

	/* set safe PATH (required for shell scripts such as /etc/rc.d/netif) */
	setenv("PATH", PATH, 1);

	/* loop reading commands */
	while (fgets(line, sizeof(line), stdin) != NULL) {
		chop(line);
		if (strncmp(line, "password ", 9) == 0) {
			char *			pass;
			char *			su_pass;

			pass = strdup(index(line, ' ') + 1);
			/* use getpwuid(0) here rather than getpwnam("root") in case "toor" or
			** other super-user name is being used */
			su_pass = strdup(getpwuid(0)->pw_passwd);
			if (strcmp(crypt(pass, su_pass), su_pass) != 0)
				printf("FAIL %s\n", gettext("invalid password"));
			else {
				auth = 1;
				printf("OK\n");
			}
			continue;
		}

		/* go no further unless authorization passed */
		if (!auth) {
			printf("FAIL %s\n", gettext("not authorized"));
			continue;
		}

		/* flag for commands that produce stdout */
		copy_stdout = 0;

		if (strcmp(line, "cat_netfile") == 0) {
			if (eaccess(NETWORKS_FILE, F_OK) == 0)
				sprintf(cmd, "%s %s", PATH_CAT, NETWORKS_FILE);
			else
				sprintf(cmd, "");
			copy_stdout = 1;
		}
		else if (strncmp(line, "write_netfile ", 14) == 0) {
			arg = index(line, ' ') + 1;
			sprintf(cmd, "%s %s %s", PATH_CP, arg, NETWORKS_FILE);
		}
		else if (strcmp(line, "diff_netfile") == 0)
			sprintf(cmd, "%s -q %s %s.save >/dev/null", PATH_DIFF, NETWORKS_FILE, NETWORKS_FILE);
		else if (strcmp(line, "backup_netfile") == 0)
			sprintf(cmd, "%s -p %s %s.save", PATH_CP, NETWORKS_FILE, NETWORKS_FILE);
		else if (strncmp(line, "restart_netif ", 14) == 0) {
			arg = index(line, ' ') + 1;
			sprintf(cmd, "%s restart %s >/dev/null 2>&1", PATH_NETIF, arg);
		}
		else if (strncmp(line, "start_netif ", 12) == 0) {
			arg = index(line, ' ') + 1;
			sprintf(cmd, "%s start %s >/dev/null 2>&1", PATH_NETIF, arg);
		}
		else if (strncmp(line, "stop_netif ", 11) == 0) {
			arg = index(line, ' ') + 1;
			sprintf(cmd, "%s stop %s >/dev/null 2>&1", PATH_NETIF, arg);
		}
		else {
			printf("FAIL %s\n", gettext("invalid command"));
			continue;
		}

		xst = system(cmd);

		if (copy_stdout)
			printf("EOF\n");

		if (WEXITSTATUS(xst) == 0)
			printf("OK\n");
		else
			printf("FAIL %s %d\n", gettext("exit status"), WEXITSTATUS(xst));
	}
}
